### PR TITLE
Provide token ttl and issue time in the audit log.

### PR DESCRIPTION
**Pull request by sgmiller**
Populate a token_ttl and token_issue_time field on the Auth struct of audit log entries, and in the Auth portion of a response for login
methods.

### CODE REVIEW

**Comment by ncabatoff** [comment]
Curious why you left out the other LeaseOptions fields?

File: audit/format.go [hunk 1]
```diff
@@ -212,6 +218,11 @@ func (f *AuditFormatter) FormatResponse(ctx context.Context, w io.Writer, config
 			NumUses:                   resp.Auth.NumUses,
 			EntityID:                  resp.Auth.EntityID,
 			TokenType:                 resp.Auth.TokenType.String(),
+			TokenTTL:                  int64(resp.Auth.TTL.Seconds()),
+		}
+		var zt time.Time
+		if resp.Auth.IssueTime != zt {
+			respAuth.TokenIssueTime = resp.Auth.IssueTime.Format(time.RFC3339)
```
Comment (mgritter) [inline audit/format.go:225]: I'm not sure why this is useful to include? Does it differ significantly from the time on the response itself?

Comment (mgritter) [reply audit/format.go:225]: Hm, could be useful for renew calls, I guess?

Comment (sgmiller) [reply audit/format.go:225]: On a login I guess it would be a similar time.  I was just trying to add it consistently.

**Comment by sgmiller** [comment]
> Curious why you left out the other LeaseOptions fields?

The remaining fields are MaxTTL (which is sort of irrelevant to an active token which has an actual ttl), Renewable, which proabably could be useful, and Increment, which doesn't apply to a token but only a renew request.

**Comment by sgmiller** [comment]
Actually, I take that back, Renewable isn't available from the TokenEntry data we get during checkToken, so we can't populate it.

File: audit/format.go [hunk 2]
```diff
@@ -4,6 +4,8 @@ import (
 	"context"
 	"crypto/tls"
 	"fmt"
+	"github.com/hashicorp/vault/sdk/helper/salt"
```
Comment (ncabatoff) [inline audit/format.go:7]: Can you revert this rearrangement of imports please?

File: audit/format.go [hunk 3]
```diff
@@ -127,6 +128,11 @@ func (f *AuditFormatter) FormatRequest(ctx context.Context, w io.Writer, config
 		},
 	}
 
+	var zt time.Time
+	if auth.IssueTime != zt {
```
Comment (ncabatoff) [inline audit/format.go:132]: I used to do this all the time, but I think you can now use the `.IsZero` method instead.

File: audit/format.go [hunk 4]
```diff
@@ -127,6 +128,11 @@ func (f *AuditFormatter) FormatRequest(ctx context.Context, w io.Writer, config
...
+		reqEntry.Auth.TokenIssueTime = auth.IssueTime.Format(time.RFC3339)
```
Comment (ncabatoff) [inline audit/format.go:133]: Is there precedent for stringifying times before auditing?  Does it work if you don't?  I'd prefer to defer formatting as late as possible.  This also means if we decide that times don't need to be redacted, it's easy to not to do so.

Comment (ncabatoff) [reply audit/format.go:133]: Sorry, I should clarify: I know we do it for some "envelope" fields like AuditResponseEntry.Time, I mean for the actual Auth/Request/Response fields that go through HashAuth/HashRequest/HashResponse.  But it looks like you're doing this after redaction, so it's moot.  Which is a bit questionable, mind you.  It's probably ok, and seems to achieve the desired effect. but it's a risky practice to get into.

**Comment by ncabatoff** [comment]
Can we update the formatter tests please, e.g. TestFormatJSONx_formatRequest?

File: audit/format.go (hunk 3 above)
Comment (sgmiller) [reply audit/format.go:132]: ah, much better.  Was not liking that construct.

File: audit/format.go (hunk 4 above)
Comment (sgmiller) [reply audit/format.go:132]: My concern was with parseability by Splunk and other systems.  I assume unix time would also work but it makes the audit log not human readable, a problem I didn't want to make worse.  Re doing after redaction, that's probably just an oversight on my part, but these fields if redacted are pretty much useless.  In particular, due to the small domain size of ttl, its pretty easy to reverse the hmac via a dictionary if you have the ability to create tokens anyway.

**Comment by sgmiller** [comment]
Will update the tests...

File: audit/format.go (hunk 4 above)
Comment (ncabatoff) [reply audit/format.go:132]: Oh, I agree that we probably don't want these fields redacted.  I just want redaction (and the decision as to whether to redact) to be in the same place as much as possible.  I also want times rendered with RFC3339, as you were doing.

**Comment by sgmiller** [comment]
Is there generalized time formatting for values in the audit log that I 
should have been using?  I'll definitely see if I can move those value 
assignments to pre-redact.

**Pull request merged**